packages/plugin-e2e/src/models/panels/visualizations.ts

```diff
@@ -0,0 +1,96 @@
+import { Locator } from '@playwright/test';
+import { PluginTestCtx, VisualizationType } from '../../types';
+
+export const visualizationDataExtractorFactory = (visualizationType: VisualizationType) => {
+  switch (visualizationType) {
+    case 'Table':
+      return new Table();
+    case 'Time series':
+      return new TimeSeries();
+    case 'Stat':
+      return new Stat();
+    case 'Gauge':
+      return new Gauge();
+    case 'Bar gauge':
+      return new BarGauge();
+    default:
+      console.log(`No visualization data extractor found for visualization type: ${visualizationType}`);
+      return new VisualizationDataExtractor(visualizationType);
+  }
+};
+
+export class VisualizationDataExtractor {
+  ctx: PluginTestCtx;
+  panel: Locator;
+
+  /**
+   * @param visualizationType the name of the visualization as it appears in the visualization picker
+   */
+  constructor(public readonly visualizationType: VisualizationType | string) {}
+
+  init(ctx: PluginTestCtx, panel: Locator) {
+    this.ctx = ctx;
+    this.panel = panel;
+  }
+
+  getValues(): Locator {
+    throw new Error(`getValues locator method was not implemented for the ${this.visualizationType} visualization`);
+  }
+  getLabels(): Locator {
+    throw new Error(`getLabels locator method was not implemented for the ${this.visualizationType} visualization`);
+  }
+}
+
+export class Table extends VisualizationDataExtractor {
+  constructor() {
+    super('Table');
+  }
+
+  getLabels(): Locator {
+    return this.panel.locator('[role="columnheader"]');
```

Comment:
Of course we'd need to add proper e2e selectors to all visualizations in Grafana an use them instead of week magic string like this (the same goes for all following extractors).